sch_cake.c

```diff
@@ -1009,13 +1010,25 @@ static bool cake_tcph_is_sack(const struct tcphdr *tcph)
 		opsize = *ptr++;
 		if (opsize < 2 || opsize > length)
 			break;
-		if (opcode == TCPOPT_SACK)
-			return true;
+		if (opcode == TCPOPT_SACK) {
+			/* ref RFC2018 for SACK format
+			 * loop to find the largest right edge value in the SACK
+			 */
+			opsize -= 2;
+			while (opsize >= 8) {
+				u32 seq;
+				seq = ntohl(*((u32 *)ptr + 1));
+				max_seq = max(max_seq, seq);
+				opsize -= 8;
+				ptr += 8;
+			}
+			break;
```

Comment:
Are all sacks guaranteed to be in the same option? This won't find any subsequent ones...

Reply:
They’re all in the same option. The SACK option is variable length, 2+8N bytes where N is the number of ‘blocks’ acknowledged.